File: 845.longest-mountain-in-array.java
import java.util.ArrayList;

/*
 * @lc app=leetcode id=845 lang=java
 *
 * [845] Longest Mountain in Array
 */

// @lc code=start
class Solution {
    // public int longestMountain(int[] arr) {
    // if (arr.length < 3) {
    // return 0;
    // }
    // int maxlen = 0;
    // int left = 0, right = 0;

    // while (left < arr.length - 1) {
    // right = left;
    // if (right < arr.length - 1 && arr[right] < arr[right + 1]) {
    // while (right < arr.length - 1 && arr[right] < arr[right + 1])
    // right++;

    // if (right < arr.length - 1 && arr[right] > arr[right + 1]) {
    // // it is a peak
    // while (right < arr.length - 1 && arr[right] > arr[right + 1])
    // right++;

    // maxlen = Math.max(maxlen, right - left + 1);
    // }
    // }

    // left = Math.max(right, left + 1);
    // }

    // return maxlen;
    // }
    public int longestMountain(int[] A) {
        int res = 0, up = 0, down = 0;
        for (int i = 1; i < A.length; ++i) {
            if (down > 0 && A[i - 1] < A[i] || A[i - 1] == A[i])
                up = down = 0;
            if (A[i - 1] < A[i])
                up++;
            if (A[i - 1] > A[i])
                down++;
            if (up > 0 && down > 0 && up + down + 1 > res)
                res = up + down + 1;
        }
        return res;
    }
}
// @lc code=end

File: 845.longest-mountain-in-array.py
#
# @lc app=leetcode id=845 lang=python3
#
# [845] Longest Mountain in Array
#

# @lc code=start
class Solution(object):
    def longestMountain(self, A):
        n = len(A)
        maxLen = start = 0

        while start < n:
            end = start
            if end + 1 < n and A[end] < A[end + 1]: #if start is a left-boundary
                #set end to the peak of this potential mountain
                while end + 1 < n and A[end] < A[end + 1]:
                    end += 1

                if end + 1 < n and A[end] > A[end + 1]: #if end is really a peak..
                    #set 'end' to right-boundary of mountain
                    while end + 1 < n and A[end] > A[end + 1]:
                        end += 1
                    #record candidate answer
                    maxLen = max(maxLen, end - start + 1)

            start = max(end, start + 1)

        return maxLen
# @lc code=end

